Infinitespark 96/issue 2309 add filtering sorting statistics to llm count

infinitespark-96 opened the pull request:
## Important note
**This branch was created from this one since I needed the new changes for sorting I recently made for span count:
infinitespark-96/issue-2234-add-sorting-to-all-columns-in-traces-view
There is an open Pull Request for this: https://github.com/comet-ml/opik/pull/2424**
## Details
In this PR I added filtering, and sorting for llm_span_count.
Also, I added statistics (average) for this field
Demo: 
![image](https://github.com/user-attachments/assets/bd06210a-7f24-408a-9b2c-d6586f38008b)
![image](https://github.com/user-attachments/assets/2d32e7df-cbe6-48f6-a3e4-3c62f2866d02)
Screencast: https://github.com/user-attachments/assets/baf16b81-30c3-4bcc-b85b-de1d9d9cb696
## Testing
**Added test coverage for sorting by llm_span_count in TracesResourceTest:**
- Extended getTracesByProject__whenSortingByValidFields__thenReturnTracesSorted to support llm_span_count
**Added test coverage for filtering by llm_span_count in TracesResourceTest:**
- whenFilterLlmSpanCountGreaterThanEqual__thenReturnTracesFiltered
- whenFilterLlmSpanCountLessThanEqual__thenReturnTracesFiltered
- whenFilterLlmSpanCountEqual__thenReturnTracesFiltered
**Added test coverage for average llm_span_count in StatsMapperTest**
- Extended mapProjectStats_returnsTotals test to include new average

andrescrz commented:
@infinitespark-96 the other PR was merged. Please fix conflicts by moving this one on top.

andrescrz commented:
> Thanks a lot for the review @andrescrz, I should have updated you on this, it's not the final version since after I reviewed your commits, which clearly show me the correct way of doing things, I realized what I should improve and change in my code. Most of your comments just confirmed that my thoughts were correct. I'll update the code in a few hours.

@infinitespark-96 thanks for the clarification. I will wait for the next revision then. Cheers.

infinitespark-96 commented on apps/opik-backend/src/main/java/com/comet/opik/domain/TraceDAO.java:
```diff
@@ -325,7 +325,7 @@ ORDER BY (workspace_id, project_id, id) DESC, last_updated_at DESC
                 sum(s.total_estimated_cost) as total_estimated_cost,
                 COUNT(s.id) AS span_count,
-                countIf(s.type = 'llm') AS llm_span_count,
+                toInt64(countIf(s.type = 'llm')) AS llm_span_count,
```
If we don't have this conversion and pass a negative value from the frontend, we get a 500 error from the server, and the actual error is:
![image](https://github.com/user-attachments/assets/801de33f-b59e-46dd-8d46-9ebdd522672a)
![image](https://github.com/user-attachments/assets/193e9993-ceab-41ec-8207-b52858308307)

I understand that maybe it's better to make the change on some other level of the backend app to prevent 500 error in cases like that (negative value from the frontend passed), as well with the UI/UX improvement to allow only positive values for this field in filters modal input since logically we cannot have negative amount of spans.
What do you think? @andrescrz

andrescrz replied:
It's fine to keep it in the meantime.

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/traces/TraceAssertions.java:
```diff
@@ -138,11 +138,19 @@ private static void assertIgnoredFields(Trace actualTrace, Trace expectedTrace,
+        expectedMap.forEach((name, expected) -> assertThat(actualMap.get(name))
                 .usingRecursiveComparison(StatsUtils.getRecursiveComparisonConfiguration())
-                .isEqualTo(expectedStats);
+                .isEqualTo(expected));
```
You're absolutely correct; I've reverted this change. Currently, I'm facing a problem, and I'm unsure of the best approach to take.
After adding an average for llmSpansCount, and making these changes int  he  StatsUtils.java file:
![image](https://github.com/user-attachments/assets/87869e61-bc02-4809-b33e-271ec031decd)
![image](https://github.com/user-attachments/assets/344aa875-f752-4dcc-9d88-4037ec71cccf)
![image](https://github.com/user-attachments/assets/406c83c4-8175-4714-a170-38ba56018e6f)

I'm having lots of tests failing because of one reason: when creating traces and don't specify the actual value of llm spans that we create for a particular trace, e.g., .llmSpanCount(0), when no actual spans are created.
 All fail on stats endpoint:
![image](https://github.com/user-attachments/assets/9fc53053-d376-4d74-805c-b6ca192412e3)
Error is the same in all the failed tests:
Actual amount of spans are 0, and average is 0 is returned, but in the expected value we have value=1.53735375E8 since llmSpanCount was auto-generated by Podam, and we don't do the real cycle of putting in the db, getting the actual page, and the traces from there with correct values for spanCount and llmSpanCount.
![image](https://github.com/user-attachments/assets/e3baf042-899b-4784-b768-47e95c29797d)
And if we manually add these correct values, all good, no error:
![image](https://github.com/user-attachments/assets/24a45472-f6fe-4333-841b-e46d32486672)
I'm a bit confused, what would be the best strategy here.. 
1. Set manually in each failed test the correct value for llmSpanCount
2. Or maybe create/modify some existing helper method and use it for trace creation since in many tests we have the repeated code for setting these values: 
![image](https://github.com/user-attachments/assets/6e66f48a-3062-47f9-bc03-c2324af0fd44)
3. Or maybe something better?
Can you help, please? @andrescrz

infinitespark-96 replied:
You can check my latest commit as a reference to what I mean here

infinitespark-96 replied:
Never mind this one, I already handled it.

infinitespark-96 commented:
> > Thanks a lot for the review @andrescrz, I should have updated you on this, it's not the final version since after I reviewed your commits, which clearly show me the correct way of doing things, I realized what I should improve and change in my code. Most of your comments just confirmed that my thoughts were correct. I'll update the code in a few hours.
> 
> @infinitespark-96 thanks for the clarification. I will wait for the next revision then. Cheers.

The latest version of the code has been pushed with all the required improvements and fixes.

In my latest commit, I addressed issues related to the average llmSpanCount in the stats endpoint. Specifically, I set:
```
.llmSpanCount(0)
.spanCount(0)
```
I’ll need to make sure the correct llmSpanCount is set everywhere to resolve the remaining failed tests related to the stats endpoint, as described in my previous question.

Other than those outstanding updates, the changes are ready for review again.
Any feedback or help would be appreciated—thanks in advance!
@andrescrz

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+            }
+
+            int llmCountUnexpected = 3;
+
```
Done

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+            int llmSpanCount = 2;
+            List<Span> spans = new ArrayList<>();
+
+            for (int i = 0; i < llmSpanCount; i++) {
```
Done

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+                        .usage(null)
+                        .totalEstimatedCost(null)
+                        .build());
+                spans.add(factory.manufacturePojo(Span.class).toBuilder()
```
Changed it, it's implemented differently now

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+                        .build());
+            }
+
+            for (Trace trace : traces.subList(1, traces.size())) {
```
Done

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+            int llmSpanCount = 1;
+            List<Span> spans = new ArrayList<>();
+
+            for (int i = 0; i < llmSpanCount; i++) {
```
Done

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+                        .build());
+            }
+
+            for (Trace trace : traces.subList(1, traces.size())) {
```
Fixed

infinitespark-96 commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java:
```diff
@@ -2566,6 +2566,213 @@ void whenFilterTotalEstimatedCostEqual_NotEqual__thenReturnTracesFiltered(String
+            int llmSpanCount = 3;
+            List<Span> spans = new ArrayList<>();
+
+            for (int i = 0; i < llmSpanCount; i++) {
```
Done
